website/content/docs/rfcs/standby-nodes-handle-read-requests.mdx

```diff
@@ -0,0 +1,237 @@
+Inside `fsm.Apply(log *raft.Log)`, the `log.Data` (which contains a `physical.StorageEntry`) is decoded. The actual operation (such as Put or Delete) is then performed on the local storage backend. This process ensures all nodes apply the same changes in the same order, keeping the state consistent.
+The methods `fsm.Snapshot()` and `fsm.Restore()` manage state snapshots. Snapshots help with log compaction and allow new or lagging nodes to catch up without replaying the entire log history.
+  - `logical.ErrStandbyPleaseForward` - For forwarding requests to leader
+  - **Remove**: `logical.ErrPerfStandbyPleaseForward` (no longer needed with unified approach)
+- **Forwarding Logic**: 
+  - Mature RPC-based request forwarding in `http/handler.go`
+  - Can be simplified by removing performance standby special cases
+- **State Management**: 
+  - `core.Standby()`, `core.StandbyStates()` for node state tracking
+  - **Simplify**: Remove `core.PerfStandby()` complexity
+- **Configuration**: 
+  - **Remove**: `disable_performance_standby` (no longer relevant)
+
+
+Any request received by a non-active node is forwarded to the active node. This is managed by `handleRequestForwarding` middleware in `http/handler.go`, which checks if the node is the leader and forwards the request if it is not.
+
+**Enhanced Forwarding Logic**: 
+Forwarding should first depend on the request method. If it's write request → forward. If read request → check if standby can serve reads and handle locally. All other logic stays the same.
+
+in `http/handler.go`:
+
+forwarding should first depend on request method, if it's write request and not leader - forward. If read request - try to serve.
+
+```
+func handleRequestForwarding(core *vault.Core, handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		isLeader, leaderAddr, _, err := core.Leader()
+		if err != nil {
+			if err == vault.ErrHANotEnabled {
+				// Standalone node, serve request normally
+				handler.ServeHTTP(w, r)
+				return
+			}
+			// Some internal error occurred
+			respondError(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		if isWriteRequest(r) && !isLeader {
+			if leaderAddr == "" {
+				respondError(w, http.StatusInternalServerError, errors.New("local node not active but active cluster node not found"))
+				return
+			}
+
+			forwardRequest(core, w, r)
+			return
+		}
+
+		handler.ServeHTTP(w, r)
+	})
+}
+
+func isWriteRequest(r *http.Request) bool {
+	switch r.Method {
+	case "POST", "PUT", "DELETE":
+		return true
+	default:
+		return false
+	}
+}
+```
+
+Read Request - doesn't mean, that we are not writing to storage! 
+
+If the logical backend attempts to write to storage, the storage layer, being in read-only state, returns `logical.ErrReadOnly`. On standby nodes, this error should be translated to `logical.ErrStandbyPleaseForward`.
```

Comment:
Right, this should be the only error. I don't necessarily think we need to translate the ErrReadOnly to the latter; both should result in a forward and mean different things / come from different layers.